jax/interpreters/ad.py

```diff
@@ -548,17 +466,50 @@ def traceable(num_primals, in_tree_def, *primals_and_tangents):
   yield out_flat, tree_def
 
 
-def call_transpose(primitive, params, call_jaxpr, args, ct):
+def call_transpose(primitive, params, call_jaxpr, args, ct, _):
   all_args, in_tree_def = tree_flatten(((), args, ct))  # empty consts
   fun = lu.hashable_partial(lu.wrap_init(backward_pass), call_jaxpr)
   fun, out_tree = flatten_fun_nokwargs(fun, in_tree_def)
   params = dict(params, name=wrap_name(params['name'], 'transpose'))
   out_flat = primitive.bind(fun, *all_args, **params)
   return tree_unflatten(out_tree(), out_flat)
 primitive_transposes[core.call_p] = partial(call_transpose, call_p)
-primitive_transposes[pe.remat_call_p] = partial(call_transpose, pe.remat_call_p)
 
-def map_transpose(primitive, params, call_jaxpr, args, ct):
+
+def remat_transpose(params, call_jaxpr, primals_in, cotangents_in, cotangent_in_avals):
+  # backward_pass can only transpose linear computations, but the call_jaxpr embedded in
+  # remat contains primal (non-linear) equations too. Hence, we have to eliminate those
+  # (in this case via partial_eval) before we call into backward_pass again.
+  typed_call_jaxpr = core.TypedJaxpr(
+      call_jaxpr, [],
+      [raise_to_shaped(p.aval if is_undefined_primal(p) else get_aval(p)) for p in primals_in],
+      cotangent_in_avals)
+  primal_jaxpr, tangent_jaxpr, out_unknowns = \
+    pe.partial_eval_jaxpr(typed_call_jaxpr,
+                          unknowns=map(is_undefined_primal, primals_in),
+                          instantiate=True,
+                          trace_type=None)
+
+  # Make a primal_jaxpr into a function. The flattening business is necessary so
+  # that we can thread UndefinedPrimals through the bind call. Note that jaxpr_as_fun
+  # doesn't understand how to handle those either, but tangent arguments are usued,
+  # so it should just ignore them.
+  primal_fun = lu.wrap_init(core.jaxpr_as_fun(primal_jaxpr))
+  flat_primals, in_tree_def = tree_flatten(primals_in)
+  primal_fun, _ = flatten_fun_nokwargs(primal_fun, in_tree_def)
+
+  params = dict(params, name=wrap_name(params['name'], 'transpose'))
+  residuals = pe.remat_call_p.bind(primal_fun, *flat_primals, **params)[len(cotangents_in):]
```

Comment:
One thing that I can't fully decide on is whether the rematerialization (of residuals, not primals!) should get wrapped in a remat too. Any opinions?

Reply:
Hmm. When would it make a difference? Two possibilities: higher-order AD, and in the way we foil XLA CSE in the translation rule of remat_call.

Maybe we should look at how the jaxprs come out when we do higher-order reverse-mode?

Reply:
Here's a linearized jaxpr on this branch:

```python
```

```
import jax
import jax.numpy as jnp

@jax.remat
def sin(x):
  return jnp.sin(x)

print(jax.make_jaxpr(lambda x, y: jax.linearize(sin, x)[1](y))(1., 2.))
```

```
{ lambda  ; a b.
  let c d = remat_call[ call_jaxpr={ lambda  ; a b.
                                     let c = cos a
                                         d = mul b c
                                     in (*, d) }
                        concrete=False
                        name=pe(pe(jvp(sin))) ] a b
  in (d,) }
```

That looks good! I _think_, but I'm not sure, that we'd like this to be its own transpose.

However, this transpose rule has a kind of "unzipping" effect (our favorite word recently):

```python
import jax
import jax.numpy as jnp

@jax.remat
def sin(x):
  return jnp.sin(x)

print(jax.make_jaxpr(lambda x, y: jax.vjp(sin, x)[1](y))(1., 2.))
```

```
{ lambda  ; a b.
  let c d e = remat_call[ call_jaxpr={ lambda  ; a.
                                       let b = cos a
                                       in (*, *, b) }
                          concrete=False
                          name=pe(pe(jvp(sin))) ] a
      f = remat_call[ call_jaxpr={ lambda  ; a b c.
                                   let d = mul c b
                                   in (d,) }
                      concrete=False
                      name=pe(transpose(pe(jvp(sin)))) ] a e b
  in (f,)
```

Questions:
1. Do we want that first jaxpr above to be its own transpose?
2. If so, how can we adapt this transpose rule to achieve that? (I have some ideas, though we should decide what we want first, and also we can land-and-iterate.)

Reply:
Yes, my primary concern was higher-order AD, because this is really the only case when remat behaves somewhat specially. Good point about inhibiting XLA CSE. Maybe this is exactly why we should keep rematerialization inside a remat?

Reply:
Interesting point about the parity-dependent idempotence of transpose! I think that it might be difficult to make transposing twice return a zipped computation again (that would be a very non-local effect, where a transpose rule would have to pull in a completely independent equation into its body). But I also think it's not _that_ important. I would be ok with saying that the transposes start looping after one application if that makes sense? The only reason to transpose right now is to do a VJP and so the zipped form would have to get unzipped to be useful anyway.

Reply:
I adapted the above example like this:

```python
import jax
import jax.numpy as jnp

@jax.remat
def sinsin(x):
  return jnp.sin(jnp.sin(x))

def linearized(a, Ta):
  b, linearized = jax.linearize(sinsin, a)
  Tb = linearized(Ta)
  return b, Tb

print(jax.make_jaxpr(linearized)(1., 2.))
```

Here's the full jaxpr we get out:

```
{ lambda  ; a b.
  let c = sin a
      d = sin c
      e f = remat_call[ call_jaxpr={ lambda  ; a b.
                                     let c = cos a
                                         d = mul b c
                                         e = sin a
                                         f = cos e
                                         g = mul d f
                                     in (*, g) }
                        concrete=False
                        name=pe(pe(jvp(sinsin))) ] a b
  in (d, f) }
```

Looks great! Here's just the linear part of type `c -> T a --o T b`:

```
{ lambda b ; a.
  let c d = remat_call[ call_jaxpr={ lambda  ; a b.
                                     let c = cos a
                                         d = mul b c
                                         e = sin a
                                         f = cos e
                                         g = mul d f
                                     in (*, g) }
                        concrete=False
                        name=pe(pe(jvp(sinsin))) ] b a
  in (d,) }
```

I _think_ the ideal transpose of that linear part, of type `c -> CT b --o CT a`, might be something like (up to re-toposorting):

```
{ lambda b ; a.
  let c d = remat_call[ call_jaxpr={ lambda  ; a b.
                                     let c = cos a
                                         d = sin a
                                         e = cos d
                                         f = mul b e
                                         g = mul f c
                                     in (*, g) }
                        concrete=False
                        name=...) ] b a
  in (d,) }
```

In words, we still get one remat call but we have the linear part of it reversed. I think that if we then transposed again we'd be back where we started (up to re-toposorting).

I think this would look like the remat transpose rule first doing some jaxpr munging (essentially what it does now, using call_transpose or something similar) but to build a single jaxpr, and then ultimately to bind a single remat_call with that jaxpr.

WDYT? Setting aside _how_ we might do it, do you think the transpose jaxpr above looks like what we'd want, or do we want it to be different somehow?

Reply:
Yeah, that makes a lot of sense! I've just pushed a change that makes it behave very similarly to what you've outlined above. I ran a sequence of transpositions and this is the result:
```
{ lambda b ; a.
  let c = remat_call[ call_jaxpr={ lambda  ; a b.
                                   let c = sin a
                                       d = cos c
                                       e = mul b d
                                       f = cos a
                                       g = mul e f
                                   in (g,) }
                      concrete=False
                      name=pe(pe(jvp(sinsin))) ] b a
  in (c,) }
{ lambda b c ; a.
  let d = remat_call[ call_jaxpr={ lambda  ; a b c.
                                   let d = cos a
                                       e = mul c d
                                       f = sin a
                                       g = cos f
                                       h = mul e g
                                   in (h,) }
                      concrete=False
                      name=pe(pe(jvp(pe(jvp(sinsin))))) ] b c a
  in (d,) }
{ lambda b c d ; a.
  let e = remat_call[ call_jaxpr={ lambda  ; a b c d.
                                   let e = sin a
                                       f = cos e
                                       g = mul d f
                                       h = cos a
                                       i = mul g h
                                   in (i,) }
                      concrete=False
                      name=pe(pe(jvp(pe(jvp(pe(jvp(sinsin))))))) ] b c d a
  in (e,) }
{ lambda b c d e ; a.
  let f = remat_call[ call_jaxpr={ lambda  ; a b c d e.
                                   let f = cos a
                                       g = mul e f
                                       h = sin a
                                       i = cos h
                                       j = mul g i
                                   in (j,) }
                      concrete=False
                      name=pe(pe(jvp(pe(jvp(pe(jvp(pe(jvp(sinsin))))))))) ] b c d e a
  in (f,) }
```

The code I've used is as following:
```python
@jax.remat                                 
def sinsin(x):                             
    return jnp.sin(jnp.sin(x))             
                                           
def value_and_grad(x, t):                  
    y, lin = jax.linearize(g, x)           
    return lin(t)                          
                                           
def vjp_and_jaxpr(f, x, ct):               
    cell = [None]                          
    _, vjp = jax.vjp(f, x)                 
    print(jax.make_jaxpr(vjp)(ct))         
    return vjp                             
                                           
vjp1 = vjp_and_jaxpr(sinsin, 1., 2.)       
vjp2 = vjp_and_jaxpr(vjp1, 3., (4.,))      
vjp3 = vjp_and_jaxpr(vjp2, (5.,), (6.,))   
vjp4 = vjp_and_jaxpr(vjp3, (7.,), ((8.,),))
```

You can see that the result is very close to what we would expect, except for the fact that `remat_call` starts accumulating more and more unused arguments. I did some digging on why it happens like that and from what I understand this is due to the fact that (1) linearization of an already linear function does not use the "primal" argument (2) neither JVP nor PE removes those unused arguments from calls, so they start to accumulate. Note that in here this is not an issue, because they're just scalars, but it might cause more issues if those things would be large arrays. Another thing to fix I guess?

Reply:
Oh and yet another issue is that VJP has this annoying behavior of wrapping things in tuples, which you can see in my code sample. If I were to iterate this method, I would have to keep wrapping the arguments more and more. That's quite silly, but it might be a little difficult to change without breaking backward compatibility. WDYT?

Reply:
Great fix and explanation! I agree those pesky details could be revised away but probably aren't worth doing right now.